src/cli/cli_tcp.cpp

```diff
@@ -0,0 +1,517 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a TCP CLI tool.
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TCP_ENABLE
+
+#include "cli_tcp.hpp"
+
+#include <openthread/tcp.h>
+
+#include "cli/cli.hpp"
+#include "common/encoding.hpp"
+#include "common/timer.hpp"
+
+namespace ot {
+namespace Cli {
+
+constexpr TcpExample::Command TcpExample::sCommands[];
+
+TcpExample::TcpExample(Interpreter &aInterpreter)
+    : mInterpreter(aInterpreter)
+    , mEndpointInitialized(false)
+    , mListenerInitialized(false)
+    , mEndpointConnected(false)
+    , mSendBusy(false)
+    , mBenchmarkBytesTotal(0)
+    , mBenchmarkLinksLeft(0)
+{
+}
+
+otError TcpExample::ProcessHelp(Arg aArgs[])
+{
+    OT_UNUSED_VARIABLE(aArgs);
+
+    for (const Command &command : sCommands)
+    {
+        mInterpreter.OutputLine(command.mName);
+    }
+
+    return OT_ERROR_NONE;
+}
+
+otError TcpExample::ProcessInit(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+    size_t  receiveBufferSize;
+
+    VerifyOrExit(!mEndpointInitialized || !mListenerInitialized, error = OT_ERROR_INVALID_STATE);
+
+    if (aArgs[0].IsEmpty())
+    {
+        receiveBufferSize = sizeof(mReceiveBuffer);
+    }
+    else
+    {
+        uint32_t windowSize;
+
+        SuccessOrExit(error = aArgs[0].ParseAsUint32(windowSize));
+        VerifyOrExit(aArgs[1].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
+
+        receiveBufferSize = windowSize + ((windowSize + 7) >> 3);
+        VerifyOrExit(receiveBufferSize <= sizeof(mReceiveBuffer) && receiveBufferSize != 0,
+                     error = OT_ERROR_INVALID_ARGS);
+    }
+
+    if (!mEndpointInitialized)
+    {
+        otTcpEndpointInitializeArgs endpointArgs;
+
+        memset(&endpointArgs, 0x00, sizeof(endpointArgs));
+        endpointArgs.mEstablishedCallback      = HandleTcpEstablishedCallback;
+        endpointArgs.mSendDoneCallback         = HandleTcpSendDoneCallback;
+        endpointArgs.mReceiveAvailableCallback = HandleTcpReceiveAvailableCallback;
+        endpointArgs.mDisconnectedCallback     = HandleTcpDisconnectedCallback;
+        endpointArgs.mContext                  = this;
+        endpointArgs.mReceiveBuffer            = mReceiveBuffer;
+        endpointArgs.mReceiveBufferSize        = receiveBufferSize;
+
+        SuccessOrExit(error = otTcpEndpointInitialize(mInterpreter.mInstance, &mEndpoint, &endpointArgs));
+        mEndpointInitialized = true;
+    }
+
+    if (!mListenerInitialized)
```

Comment:
Back when I was writing this, I think that I was thinking of the case where initialization succeeds for one but not the other. But in retrospect, that is better handled by deinitializing one if initializing the other fails. I'll go ahead and implement that, and also switch to using one bool instead of two.

Reply:
Deinitializing sounds good to me.

But I am wonder what might possibility go wrong in `otTcpListenerInitialize` or `otTcpEndpointInitialize`. They should always succeed?

Reply:
They *should* always succeed, But that's based on looking at the implementation. The API still has a return code that the CLI tool should handle as well (since someone may later change the implementation so that there's a chance of failure...)